<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.429">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Data Science for Biodiversity Scientists - References</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": true,
  "collapse-after": 1,
  "panel-placement": "start",
  "type": "overlay",
  "limit": 10,
  "keyboard-shortcut": [
    null
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./references.html">References</a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Data Science for Biodiversity Scientists</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="Download" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download"><i class="bi bi-download"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Data-Science-for-Biodiversity-Scientists.pdf">
              <i class="bi bi-bi-file-pdf pe-1"></i>
            Download PDF
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Data-Science-for-Biodiversity-Scientists.epub">
              <i class="bi bi-bi-journal pe-1"></i>
            Download ePub
            </a>
          </li>
      </ul>
    </div>
    <div id="quarto-search" class="quarto-navigation-tool px-1" title="Search"></div>
</div>
    </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapters/clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Clustering</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapters/gradientdescent.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Gradient descent</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapters/crossvalidation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Cross-validation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapters/classification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Supervised classification</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapters/variableselection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Variable preparation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapters/learningcurves.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Hyper-parameters tuning</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapters/explanations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Explaining predictions</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">References</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-allouche2006" class="csl-entry" role="listitem">
ALLOUCHE, O., TSOAR, A. &amp; KADMON, R. (2006). <a href="https://doi.org/10.1111/j.1365-2664.2006.01214.x">Assessing the
accuracy of species distribution models: prevalence, kappa and the true
skill statistic (TSS)</a>. <em>Journal of Applied Ecology</em>, 43,
1223–1232.
</div>
<div id="ref-aloise2009" class="csl-entry" role="listitem">
Aloise, D., Deshpande, A., Hansen, P. &amp; Popat, P. (2009). <a href="https://doi.org/10.1007/s10994-009-5103-0">NP-hardness of
Euclidean sum-of-squares clustering</a>. <em>Machine Learning</em>, 75,
245–248.
</div>
<div id="ref-amarasinghe2023" class="csl-entry" role="listitem">
Amarasinghe, K., Rodolfa, K.T., Lamba, H. &amp; Ghani, R. (2023). <a href="https://doi.org/10.1017/dap.2023.2">Explainable machine learning
for public policy: Use cases, gaps, and research directions</a>.
<em>Data &amp; Policy</em>, 5.
</div>
<div id="ref-anderson1928" class="csl-entry" role="listitem">
Anderson, E. (1928). <a href="https://doi.org/10.2307/2394087">The
problem of species in the northern blue flags, iris versicolor l. And
iris virginica l.</a> <em>Annals of the Missouri Botanical Garden</em>,
15, 241.
</div>
<div id="ref-aono2008" class="csl-entry" role="listitem">
Aono, Y. &amp; Kazui, K. (2008). <a href="https://doi.org/10.1002/joc.1594">Phenological data series of
cherry tree flowering in Kyoto, Japan, and its application to
reconstruction of springtime temperatures since the 9th century</a>.
<em>International Journal of Climatology</em>, 28, 905–914.
</div>
<div id="ref-aono2009" class="csl-entry" role="listitem">
Aono, Y. &amp; Saito, S. (2009). <a href="https://doi.org/10.1007/s00484-009-0272-x">Clarifying springtime
temperature reconstructions of the medieval period by gap-filling the
cherry blossom phenological data series at Kyoto, Japan</a>.
<em>International Journal of Biometeorology</em>, 54, 211–219.
</div>
<div id="ref-barbet-massin2011" class="csl-entry" role="listitem">
Barbet-Massin, M. &amp; Jiguet, F. (2011). <a href="https://doi.org/10.1371/journal.pone.0018228">Back from a
Predicted Climatic Extinction of an Island Endemic: A Future for the
Corsican Nuthatch</a>. <em>PLoS ONE</em>, 6, e18228.
</div>
<div id="ref-barbet-massin2012" class="csl-entry" role="listitem">
Barbet-Massin, M., Jiguet, F., Albert, C.H. &amp; Thuiller, W. (2012).
<a href="https://doi.org/10.1111/j.2041-210x.2011.00172.x">Selecting
pseudo-absences for species distribution models: how, where and how
many?</a> <em>Methods in Ecology and Evolution</em>, 3, 327–338.
</div>
<div id="ref-beery2021" class="csl-entry" role="listitem">
Beery, S., Cole, E., Parker, J., Perona, P. &amp; Winner, K. (2021). <a href="https://doi.org/10.1145/3460112.3471966">Species distribution
modeling for machine learning practitioners: A review</a>. <em>ACM
SIGCAS Conference on Computing and Sustainable Societies (COMPASS)</em>.
</div>
<div id="ref-bell2022" class="csl-entry" role="listitem">
Bell, A., Solano-Kamaiko, I., Nov, O. &amp; Stoyanovich, J. (2022). <a href="https://doi.org/10.1145/3531146.3533090">It<span>’</span>s just
not that simple: An empirical study of the accuracy-explainability
trade-off in machine learning for public policy</a>. <em>2022 ACM
Conference on Fairness, Accountability, and Transparency</em>.
</div>
<div id="ref-bellocchi2010" class="csl-entry" role="listitem">
Bellocchi, G., Rivington, M., Donatelli, M. &amp; Matthews, K. (2010).
<a href="https://doi.org/10.1051/agro/2009001">Validation of biophysical
models: issues and methodologies. A review</a>. <em>Agronomy for
Sustainable Development</em>, 30, 109–130.
</div>
<div id="ref-bergmeir2012" class="csl-entry" role="listitem">
Bergmeir, C. &amp; Benítez, J.M. (2012). <a href="https://doi.org/10.1016/j.ins.2011.12.028">On the use of
cross-validation for time series predictor evaluation</a>.
<em>Information Sciences</em>, 191, 192–213.
</div>
<div id="ref-berteaux2014" class="csl-entry" role="listitem">
Berteaux, D. (2014). <em><a href="https://doi.org/10.1353/book35753">Changements climatiques et
biodiversité du québec</a></em>. Presses de l’Université du Québec.
</div>
<div id="ref-bezanson2017" class="csl-entry" role="listitem">
Bezanson, J., Edelman, A., Karpinski, S. &amp; Shah, V.B. (2017). <a href="https://doi.org/10.1137/141000671">Julia: A Fresh Approach to
Numerical Computing</a>. <em>SIAM Review</em>, 59, 65–98.
</div>
<div id="ref-blaom2020" class="csl-entry" role="listitem">
Blaom, A., Kiraly, F., Lienart, T., Simillides, Y., Arenas, D. &amp;
Vollmer, S. (2020). <a href="https://doi.org/10.21105/joss.02704">MLJ: A
julia package for composable machine learning</a>. <em>Journal of Open
Source Software</em>, 5, 2704.
</div>
<div id="ref-bodmer2021" class="csl-entry" role="listitem">
Bodmer, W., Bailey, R.A., Charlesworth, B., Eyre-Walker, A., Farewell,
V., Mead, A., <em>et al.</em> (2021). <a href="https://doi.org/10.1038/s41437-020-00394-6">The outstanding
scientist, R.A. Fisher: his views on eugenics and race</a>.
<em>Heredity</em>, 126, 565–576.
</div>
<div id="ref-booth2022" class="csl-entry" role="listitem">
Booth, T.H. (2022). <a href="https://doi.org/10.1111/aec.13234">Checking
bioclimatic variables that combine temperature and precipitation data
before their use in species distribution models</a>. <em>Austral
Ecology</em>, 47, 1506–1514.
</div>
<div id="ref-brose2004" class="csl-entry" role="listitem">
Brose, U., Ostling, A., Harrison, K. &amp; Martinez, N.D. (2004). <a href="https://doi.org/10.1038/nature02297">Unified spatial scaling of
species and their trophic interactions</a>. <em>Nature</em>, 428,
167–171.
</div>
<div id="ref-celisse2014" class="csl-entry" role="listitem">
Celisse, A. (2014). <a href="https://doi.org/10.1214/14-aos1240">Optimal
cross-validation in density estimation with the
<span>$</span>l<span>^</span><span></span>2<span></span><span>$</span>-loss</a>.
<em>The Annals of Statistics</em>, 42.
</div>
<div id="ref-chicco2020" class="csl-entry" role="listitem">
Chicco, D. &amp; Jurman, G. (2020). <a href="https://doi.org/10.1186/s12864-019-6413-7">The advantages of the
Matthews correlation coefficient (MCC) over F1 score and accuracy in
binary classification evaluation</a>. <em>BMC Genomics</em>, 21.
</div>
<div id="ref-chicco2023" class="csl-entry" role="listitem">
Chicco, D. &amp; Jurman, G. (2023). <a href="https://doi.org/10.1186/s13040-023-00322-4">The Matthews
correlation coefficient&nbsp;(MCC) should replace the ROC&nbsp;AUC as the standard
metric for assessing binary classification</a>. <em>BioData Mining</em>,
16.
</div>
<div id="ref-cholletramampiandra2023" class="csl-entry" role="listitem">
Chollet Ramampiandra, E., Scheidegger, A., Wydler, J. &amp; Schuwirth,
N. (2023). <a href="https://doi.org/10.1016/j.ecolmodel.2023.110353">A
comparison of machine learning and statistical species distribution
models: Quantifying overfitting supports model interpretation</a>.
<em>Ecological Modelling</em>, 481, 110353.
</div>
<div id="ref-clapham1935" class="csl-entry" role="listitem">
Clapham, A.R., Raunkiaer, C., Gilbert-Carter, H., Tansley, A.G. &amp;
Fausboll. (1935). <a href="https://doi.org/10.2307/2256153">The life
forms of plants and statistical plant geography.</a> <em>The Journal of
Ecology</em>, 23, 247.
</div>
<div id="ref-cohen1984" class="csl-entry" role="listitem">
Cohen, J.E. &amp; Briand, F. (1984). Trophic links of community food
webs. <em>Proc Natl Acad Sci U S A</em>, 81, 4105–4109.
</div>
<div id="ref-cooney2022" class="csl-entry" role="listitem">
Cooney, C.R., He, Y., Varley, Z.K., Nouri, L.O., Moody, C.J.A., Jardine,
M.D., <em>et al.</em> (2022). <a href="https://doi.org/10.1038/s41559-022-01714-1">Latitudinal gradients
in avian colourfulness</a>. <em>Nature Ecology &amp; Evolution</em>, 6,
622–629.
</div>
<div id="ref-cooper2019" class="csl-entry" role="listitem">
Cooper, N., Bond, A.L., Davis, J.L., Portela Miguez, R., Tomsett, L.
&amp; Helgen, K.M. (2019). <a href="https://doi.org/10.1098/rspb.2019.2025">Sex biases in bird and
mammal natural history collections</a>. <em>Proceedings of the Royal
Society B: Biological Sciences</em>, 286, 20192025.
</div>
<div id="ref-davies1979" class="csl-entry" role="listitem">
Davies, D.L. &amp; Bouldin, D.W. (1979). <a href="https://doi.org/10.1109/tpami.1979.4766909">A cluster separation
measure</a>. <em>IEEE Transactions on Pattern Analysis and Machine
Intelligence</em>, PAMI-1, 224–227.
</div>
<div id="ref-demarco2018" class="csl-entry" role="listitem">
De Marco, P. &amp; Nóbrega, C.C. (2018). <a href="https://doi.org/10.1371/journal.pone.0202403">Evaluating
collinearity effects on species distribution models: An approach based
on virtual species simulation</a>. <em>PLOS ONE</em>, 13, e0202403.
</div>
<div id="ref-deisenroth2020" class="csl-entry" role="listitem">
Deisenroth, M.P., Faisal, A.A. &amp; Ong, C.S. (2020). <a href="https://doi.org/10.1017/9781108679930">Mathematics for machine
learning</a>.
</div>
<div id="ref-dormann2012" class="csl-entry" role="listitem">
Dormann, C.F., Elith, J., Bacher, S., Buchmann, C., Carl, G., Carré, G.,
<em>et al.</em> (2012). <a href="https://doi.org/10.1111/j.1600-0587.2012.07348.x">Collinearity: a
review of methods to deal with it and a simulation study evaluating
their performance</a>. <em>Ecography</em>, 36, 27–46.
</div>
<div id="ref-dornelas2018" class="csl-entry" role="listitem">
Dornelas, M., Antão, L.H., Moyes, F., Bates, A.E., Magurran, A.E., Adam,
D., <em>et al.</em> (2018). <a href="https://doi.org/10.1111/geb.12729">BioTIME: A database of
biodiversity time series for the Anthropocene</a>. <em>Global Ecology
and Biogeography</em>, 27, 760–786.
</div>
<div id="ref-dunn1974" class="csl-entry" role="listitem">
Dunn, J.C. (1974). <a href="https://doi.org/10.1080/01969727408546059">Well-Separated Clusters
and Optimal Fuzzy Partitions</a>. <em>Journal of Cybernetics</em>, 4,
95–104.
</div>
<div id="ref-elith2009" class="csl-entry" role="listitem">
Elith, J. &amp; Leathwick, J.R. (2009). <a href="https://doi.org/10.1146/annurev.ecolsys.110308.120159">Species
Distribution Models: Ecological Explanation and Prediction Across Space
and Time</a>. <em>Annual Review of Ecology, Evolution, and
Systematics</em>, 40, 677–697.
</div>
<div id="ref-fawcett2006" class="csl-entry" role="listitem">
Fawcett, T. (2006). <a href="https://doi.org/10.1016/j.patrec.2005.10.010">An introduction to
ROC analysis</a>. <em>Pattern Recognition Letters</em>, 27, 861–874.
</div>
<div id="ref-fick2017" class="csl-entry" role="listitem">
Fick, S.E. &amp; Hijmans, R.J. (2017). <a href="https://doi.org/10.1002/joc.5086">WorldClim 2: new
1<span>-</span>km spatial resolution climate surfaces for global land
areas</a>. <em>International Journal of Climatology</em>, 37, 4302–4315.
</div>
<div id="ref-fisher1936" class="csl-entry" role="listitem">
Fisher, R.A. (1936). <a href="https://doi.org/10.1111/j.1469-1809.1936.tb02137.x">The Use Of
Multiple Measurements In Taxonomic Problems</a>. <em>Annals of
Eugenics</em>, 7, 179–188.
</div>
<div id="ref-fox2017" class="csl-entry" role="listitem">
Fox, E.W., Hill, R.A., Leibowitz, S.G., Olsen, A.R., Thornbrugh, D.J.
&amp; Weber, M.H. (2017). <a href="https://doi.org/10.1007/s10661-017-6025-0">Assessing the accuracy
and stability of variable selection methods for random forest modeling
in ecology</a>. <em>Environmental Monitoring and Assessment</em>, 189.
</div>
<div id="ref-gonzalez2023" class="csl-entry" role="listitem">
Gonzalez, A., Vihervaara, P., Balvanera, P., Bates, A.E., Bayraktarov,
E., Bellingham, P.J., <em>et al.</em> (2023). <a href="https://doi.org/10.1038/s41559-023-02171-0">A global biodiversity
observing system to unite monitoring and guide action</a>. <em>Nature
Ecology &amp; Evolution</em>.
</div>
<div id="ref-vandergoot2021" class="csl-entry" role="listitem">
Goot, R. van der. (2021). <a href="https://doi.org/10.18653/v1/2021.emnlp-main.368">We need to talk
about train-dev-test splits</a>. <em>Proceedings of the 2021 Conference
on Empirical Methods in Natural Language Processing</em>.
</div>
<div id="ref-gorman2014" class="csl-entry" role="listitem">
Gorman, K.B., Williams, T.D. &amp; Fraser, W.R. (2014). <a href="https://doi.org/10.1371/journal.pone.0090081">Ecological Sexual
Dimorphism and Environmental Variability within a Community of Antarctic
Penguins (Genus Pygoscelis)</a>. <em>PLoS ONE</em>, 9, e90081.
</div>
<div id="ref-graham2003" class="csl-entry" role="listitem">
Graham, M.H. (2003). <a href="https://doi.org/10.1890/02-3114">CONFRONTING MULTICOLLINEARITY IN
ECOLOGICAL MULTIPLE REGRESSION</a>. <em>Ecology</em>, 84, 2809–2815.
</div>
<div id="ref-gray1984" class="csl-entry" role="listitem">
Gray, R. (1984). <a href="https://doi.org/10.1109/massp.1984.1162229">Vector
quantization</a>. <em>IEEE ASSP Magazine</em>, 1, 4–29.
</div>
<div id="ref-guillera-arroita2015" class="csl-entry" role="listitem">
Guillera-Arroita, G., Lahoz-Monfort, J.J., Elith, J., Gordon, A.,
Kujala, H., Lentini, P.E., <em>et al.</em> (2015). <a href="https://doi.org/10.1111/geb.12268">Is my species distribution
model fit for purpose? Matching data and models to applications</a>.
<em>Global Ecology and Biogeography</em>, 24, 276–292.
</div>
<div id="ref-hanberry2012" class="csl-entry" role="listitem">
Hanberry, B.B., He, H.S. &amp; Palik, B.J. (2012). <a href="https://doi.org/10.1371/journal.pone.0044486">Pseudoabsence
Generation Strategies for Species Distribution Models</a>. <em>PLoS
ONE</em>, 7, e44486.
</div>
<div id="ref-hand2001" class="csl-entry" role="listitem">
Hand, D.J. &amp; Yu, K. (2001). <a href="https://doi.org/10.2307/1403452">Idiot’s bayes: Not so stupid
after all?</a> <em>International Statistical Review / Revue
Internationale de Statistique</em>, 69, 385.
</div>
<div id="ref-hanley1982" class="csl-entry" role="listitem">
Hanley, J.A. &amp; McNeil, B.J. (1982). <a href="https://doi.org/10.1148/radiology.143.1.7063747">The meaning and
use of the area under a receiver operating characteristic (ROC)
curve.</a> <em>Radiology</em>, 143, 29–36.
</div>
<div id="ref-hawkins2003" class="csl-entry" role="listitem">
Hawkins, D.M., Basak, S.C. &amp; Mills, D. (2003). <a href="https://doi.org/10.1021/ci025626i">Assessing Model Fit by
Cross-Validation</a>. <em>Journal of Chemical Information and Computer
Sciences</em>, 43, 579–586.
</div>
<div id="ref-hijmans2012" class="csl-entry" role="listitem">
Hijmans, R.J. (2012). <a href="https://doi.org/10.1890/11-0826.1">Cross-validation of species
distribution models: removing spatial sorting bias and calibration with
a null model</a>. <em>Ecology</em>, 93, 679–688.
</div>
<div id="ref-homeyer2022" class="csl-entry" role="listitem">
Homeyer, A., Geißler, C., Schwen, L.O., Zakrzewski, F., Evans, T.,
Strohmenger, K., <em>et al.</em> (2022). <a href="https://doi.org/10.1038/s41379-022-01147-y">Recommendations on
compiling test datasets for evaluating artificial intelligence solutions
in pathology</a>. <em>Modern Pathology</em>, 35, 1759–1769.
</div>
<div id="ref-horst2020" class="csl-entry" role="listitem">
Horst, A.M., Hill, A.P. &amp; Gorman, K.B. (2020). <em><a href="https://doi.org/10.5281/ZENODO.3960218">Allisonhorst/palmerpenguins:
v0.1.0</a></em>. Zenodo.
</div>
<div id="ref-howley2005" class="csl-entry" role="listitem">
Howley, T., Madden, M.G., O’Connell, M.-L. &amp; Ryder, A.G. (2005). <a href="https://doi.org/10.1007/1-84628-224-1_16">The effect of principal
component analysis on machine learning accuracy with high dimensional
spectral data</a>. Springer London, pp. 209–222.
</div>
<div id="ref-innes2018" class="csl-entry" role="listitem">
Innes, M. (2018). <a href="https://doi.org/10.48550/ARXIV.1810.07951">Don’t unroll adjoint:
Differentiating SSA-form programs</a>.
</div>
<div id="ref-Jurman2012" class="csl-entry" role="listitem">
Jurman, G., Riccadonna, S. &amp; Furlanello, C. (2012). <a href="https://doi.org/10.1371/journal.pone.0041882">A Comparison of MCC
and CEN Error Measures in Multi-Class Prediction</a>. <em>PLoS ONE</em>,
7, e41882.
</div>
<div id="ref-karger2017" class="csl-entry" role="listitem">
Karger, D.N., Conrad, O., Böhner, J., Kawohl, T., Kreft, H., Soria-Auza,
R.W., <em>et al.</em> (2017). <a href="https://doi.org/10.1038/sdata.2017.122">Climatologies at high
resolution for the earth<span>’</span>s land surface areas</a>.
<em>Scientific Data</em>, 4.
</div>
<div id="ref-kaufman2011" class="csl-entry" role="listitem">
Kaufman, S., Rosset, S. &amp; Perlich, C. (2011). <a href="https://doi.org/10.1145/2020408.2020496">Leakage in data
mining</a>. <em>Proceedings of the 17th ACM SIGKDD international
conference on Knowledge discovery and data mining</em>.
</div>
<div id="ref-kennedy2020" class="csl-entry" role="listitem">
Kennedy, S. &amp; Burbach, M. (2020). <a href="https://doi.org/10.1353/gpr.2020.0016">Great Plains Ranchers
Managing for Vegetation Heterogeneity: A Multiple Case Study</a>.
<em>Great Plains Research</em>, 30, 137–148.
</div>
<div id="ref-kessy2018" class="csl-entry" role="listitem">
Kessy, A., Lewin, A. &amp; Strimmer, K. (2018). <a href="https://doi.org/10.1080/00031305.2016.1277159">Optimal Whitening
and Decorrelation</a>. <em>The American Statistician</em>, 72, 309–314.
</div>
<div id="ref-koivunen1999" class="csl-entry" role="listitem">
Koivunen, A.C. &amp; Kostinski, A.B. (1999). <a href="https://doi.org/10.1175/1520-0450(1999)038<0741:tfodwt>2.0.co;2">The
Feasibility of Data Whitening to Improve Performance of Weather
Radar</a>. <em>Journal of Applied Meteorology</em>, 38, 741–749.
</div>
<div id="ref-kosanic2020" class="csl-entry" role="listitem">
Kosanic, A. &amp; Petzold, J. (2020). <a href="https://doi.org/10.1016/j.ecoser.2020.101168">A systematic review
of cultural ecosystem services and human wellbeing</a>. <em>Ecosystem
Services</em>, 45, 101168.
</div>
<div id="ref-kupervasser2014" class="csl-entry" role="listitem">
Kupervasser, O. (2014). <a href="https://doi.org/10.1134/s1054661814010088">The mysterious
optimality of Naive Bayes: Estimation of the probability in the system
of <span>“</span>classifiers<span>”</span></a>. <em>Pattern Recognition
and Image Analysis</em>, 24, 1–10.
</div>
<div id="ref-legendre2012" class="csl-entry" role="listitem">
Legendre, P. &amp; Legendre, L. (2012). <em>Numerical ecology</em>.
Developments in environmental modelling. Third English edition.
Elsevier, Oxford, UK.
</div>
<div id="ref-leroy2018" class="csl-entry" role="listitem">
Leroy, B., Delsol, R., Hugueny, B., Meynard, C.N., Barhoumi, C.,
Barbet-Massin, M., <em>et al.</em> (2018). <a href="https://doi.org/10.1111/jbi.13402">Without quality
presence<span></span>absence data, discrimination metrics such as TSS
can be misleading measures of model performance</a>. <em>Journal of
Biogeography</em>, 45, 1994–2002.
</div>
<div id="ref-luccioni2023" class="csl-entry" role="listitem">
Luccioni, A.S. &amp; Rolnick, D. (2023). <a href="https://doi.org/10.1609/aaai.v37i12.26682">Bugs in the data: How
ImageNet misrepresents biodiversity</a>. <em>Proceedings of the AAAI
Conference on Artificial Intelligence</em>, 37, 14382–14390.
</div>
<div id="ref-lundberg2017" class="csl-entry" role="listitem">
Lundberg, S.M. &amp; Lee, S.-I. (2017). <a href="https://proceedings.neurips.cc/paper_files/paper/2017/file/8a20a8621978632d76c43dfd28b67767-Paper.pdf">A
unified approach to interpreting model predictions</a>. In: <em>Advances
in neural information processing systems</em> (eds. Guyon, I., Luxburg,
U.V., Bengio, S., Wallach, H., Fergus, R., Vishwanathan, S., et al.).
Curran Associates, Inc.
</div>
<div id="ref-macdonald2020" class="csl-entry" role="listitem">
MacDonald, A.A.M., Banville, F. &amp; Poisot, T. (2020). <a href="https://doi.org/10.1016/j.patter.2020.100079">Revisiting the
links-species scaling relationship in food webs</a>. <em>Patterns</em>,
1.
</div>
<div id="ref-martinez1992" class="csl-entry" role="listitem">
Martinez, N.D. (1992). <a href="http://www.jstor.org/stable/2462337">Constant connectance in
community food webs</a>. <em>The American Naturalist</em>, 139,
1208–1218.
</div>
<div id="ref-martinez-meyer2005" class="csl-entry" role="listitem">
Martinez-Meyer, E. (2005). <a href="https://doi.org/10.17161/bi.v2i0.8">Climate change and
biodiversity: Some considerations in forecasting shifts in species’
potential distributions</a>. <em>Biodiversity Informatics</em>, 2.
</div>
<div id="ref-mcfeeters2013" class="csl-entry" role="listitem">
McFeeters, S. (2013). <a href="https://doi.org/10.3390/rs5073544">Using
the Normalized Difference Water Index (NDWI) within a Geographic
Information System to Detect Swimming Pools for Mosquito Abatement: A
Practical Approach</a>. <em>Remote Sensing</em>, 5, 3544–3561.
</div>
<div id="ref-mesgaran2014" class="csl-entry" role="listitem">
Mesgaran, M.B., Cousens, R.D. &amp; Webber, B.L. (2014). <a href="https://doi.org/10.1111/ddi.12209">Here be dragons: a tool for
quantifying novelty due to covariate range and correlation change when
projecting species distribution models</a>. <em>Diversity and
Distributions</em>, 20, 1147–1159.
</div>
<div id="ref-mimet2009" class="csl-entry" role="listitem">
Mimet, A., Pellissier, V., Quénol, H., Aguejdad, R., Dubreuil, V. &amp;
Rozé, F. (2009). <a href="https://doi.org/10.1007/s00484-009-0214-7">Urbanisation induces
early flowering: evidence from Platanus acerifolia and Prunus
cerasus</a>. <em>International Journal of Biometeorology</em>, 53,
287–298.
</div>
<div id="ref-mondejar2019" class="csl-entry" role="listitem">
Mondejar, J.P. &amp; Tongco, A.F. (2019). <a href="https://doi.org/10.1186/s42834-019-0016-5">Near infrared band of
Landsat 8 as water index: a case study around Cordova and Lapu-Lapu
City, Cebu, Philippines</a>. <em>Sustainable Environment Research</em>,
29.
</div>
<div id="ref-moriuchi2019" class="csl-entry" role="listitem">
Moriuchi, E. &amp; Basil, M. (2019). <a href="https://doi.org/10.3390/su11061820">The Sustainability of Ohanami
Cherry Blossom Festivals as a Cultural Icon</a>.
<em>Sustainability</em>, 11, 1820.
</div>
<div id="ref-morrison1999" class="csl-entry" role="listitem">
Morrison, M. &amp; Morgan, M.S. (1999). <a href="https://doi.org/10.1017/cbo9780511660108.003">Models as mediating
instruments</a>. Cambridge University Press, pp. 10–37.
</div>
<div id="ref-murtaugh2009" class="csl-entry" role="listitem">
Murtaugh, P.A. (2009). <a href="https://doi.org/10.1111/j.1461-0248.2009.01361.x">Performance of
several variable<span>-</span>selection methods applied to real
ecological data</a>. <em>Ecology Letters</em>, 12, 1061–1068.
</div>
<div id="ref-newey2015" class="csl-entry" role="listitem">
Newey, S., Davidson, P., Nazir, S., Fairhurst, G., Verdicchio, F.,
Irvine, R.J., <em>et al.</em> (2015). <a href="https://doi.org/10.1007/s13280-015-0713-1">Limitations of
recreational camera traps for wildlife management and conservation
research: A practitioner<span>’</span>s perspective</a>. <em>Ambio</em>,
44, 624–635.
</div>
<div id="ref-nisbet2018" class="csl-entry" role="listitem">
Nisbet, R., Miner, G., Yale, K., Elder, J.F. &amp; Peterson, A.F.
(2018). <em>Handbook of statistical analysis and data mining
applications</em>. Second edition. Academic Press, London.
</div>
<div id="ref-ohashi2011a" class="csl-entry" role="listitem">
Ohashi, Y., Kawakami, H., Shigeta, Y., Ikeda, H. &amp; Yamamoto, N.
(2011). <a href="https://doi.org/10.1007/s00484-011-0496-4">The
phenology of cherry blossom (Prunus yedoensis
<span>“</span>Somei-yoshino<span>”</span>) and the geographic features
contributing to its flowering</a>. <em>International Journal of
Biometeorology</em>, 56, 903–914.
</div>
<div id="ref-pearson1901" class="csl-entry" role="listitem">
Pearson, K. (1901). <a href="https://doi.org/10.1080/14786440109462720">LIII. <span><em>On
lines and planes of closest fit to systems of points in
space</em></span></a>. <em>The London, Edinburgh, and Dublin
Philosophical Magazine and Journal of Science</em>, 2, 559–572.
</div>
<div id="ref-perkins2005" class="csl-entry" role="listitem">
Perkins, N.J. &amp; Schisterman, E.F. (2005). <a href="https://doi.org/10.1002/bimj.200410133">The Youden Index and the
Optimal Cut-Point Corrected for Measurement Error</a>. <em>Biometrical
Journal</em>, 47, 428–441.
</div>
<div id="ref-Perkins2006" class="csl-entry" role="listitem">
Perkins, N.J. &amp; Schisterman, E.F. (2006). <a href="https://doi.org/10.1093/aje/kwj063">The Inconsistency of
<span>“</span>Optimal<span>”</span> Cutpoints Obtained using Two
Criteria based on the Receiver Operating Characteristic Curve</a>.
<em>American Journal of Epidemiology</em>, 163, 670–675.
</div>
<div id="ref-perl2022" class="csl-entry" role="listitem">
Perl, R.G.B., Avidor, E., Roll, U., Malka, Y., Geffen, E. &amp; Gafny,
S. (2022). <a href="https://doi.org/10.1002/edn3.276">Using eDNA
presence/non<span>-</span>detection data to characterize the abiotic and
biotic habitat requirements of a rare, elusive amphibian</a>.
<em>Environmental DNA</em>, 4, 642–653.
</div>
<div id="ref-Peterson2018" class="csl-entry" role="listitem">
Peterson, A.T., Asase, A., Canhos, D., Souza, S. de &amp; Wieczorek, J.
(2018). <a href="https://doi.org/10.3897/bdj.6.e26826">Data leakage and
loss in biodiversity informatics</a>. <em>Biodiversity Data
Journal</em>, 6.
</div>
<div id="ref-petitpierre2016" class="csl-entry" role="listitem">
Petitpierre, B., Broennimann, O., Kueffer, C., Daehler, C. &amp; Guisan,
A. (2016). <a href="https://doi.org/10.1111/geb.12530">Selecting
predictors to maximize the transferability of species distribution
models: lessons from cross<span>-</span>continental plant invasions</a>.
<em>Global Ecology and Biogeography</em>, 26, 275–287.
</div>
<div id="ref-poisot2023a" class="csl-entry" role="listitem">
Poisot, T. (2023). <a href="https://doi.org/10.1111/2041-210x.14071">Guidelines for the
prediction of species interactions through binary classification</a>.
<em>Methods in Ecology and Evolution</em>, 14, 1333–1345.
</div>
<div id="ref-poisot2023" class="csl-entry" role="listitem">
Poisot, T., Ouellet, M.-A., Mollentze, N., Farrell, M.J., Becker, D.J.,
Brierley, L., <em>et al.</em> (2023). <a href="https://doi.org/10.1016/j.patter.2023.100738">Network embedding
unveils the hidden interactions in the mammalian virome</a>.
<em>Patterns</em>, 4, 100738.
</div>
<div id="ref-powers2020" class="csl-entry" role="listitem">
Powers, D.M.W. (2020). <a href="https://doi.org/10.48550/ARXIV.2010.16061">Evaluation: From
precision, recall and f-measure to ROC, informedness, markedness and
correlation</a>. <em>arXiv</em>.
</div>
<div id="ref-primack2009" class="csl-entry" role="listitem">
Primack, R.B., Higuchi, H. &amp; Miller-Rushing, A.J. (2009). <a href="https://doi.org/10.1016/j.biocon.2009.03.016">The impact of
climate change on cherry trees and other species in Japan</a>.
<em>Biological Conservation</em>, 142, 1943–1949.
</div>
<div id="ref-roberts2017" class="csl-entry" role="listitem">
Roberts, D.R., Bahn, V., Ciuti, S., Boyce, M.S., Elith, J.,
Guillera-Arroita, G., <em>et al.</em> (2017). <a href="https://doi.org/10.1111/ecog.02881">Cross<span>-</span>validation
strategies for data with temporal, spatial, hierarchical, or
phylogenetic structure</a>. <em>Ecography</em>, 40, 913–929.
</div>
<div id="ref-rousseeuw1987" class="csl-entry" role="listitem">
Rousseeuw, P.J. (1987). <a href="https://doi.org/10.1016/0377-0427(87)90125-7">Silhouettes: A
graphical aid to the interpretation and validation of cluster
analysis</a>. <em>Journal of Computational and Applied Mathematics</em>,
20, 53–65.
</div>
<div id="ref-roy2006" class="csl-entry" role="listitem">
Roy, D.P., Boschetti, L. &amp; Trigg, S.N. (2006). <a href="https://doi.org/10.1109/lgrs.2005.858485">Remote Sensing of Fire
Severity: Assessing the Performance of the Normalized Burn Ratio</a>.
<em>IEEE Geoscience and Remote Sensing Letters</em>, 3, 112–116.
</div>
<div id="ref-runghen2022" class="csl-entry" role="listitem">
Runghen, R., Stouffer, D.B. &amp; Dalla Riva, G.V. (2022). <a href="https://doi.org/10.1098/rsos.220079">Exploiting node metadata to
predict interactions in bipartite networks using graph embedding and
neural networks</a>. <em>Royal Society Open Science</em>, 9.
</div>
<div id="ref-sakurai2011" class="csl-entry" role="listitem">
Sakurai, R., Jacobson, S.K., Kobori, H., Primack, R., Oka, K., Komatsu,
N., <em>et al.</em> (2011). <a href="https://doi.org/10.1016/j.biocon.2010.09.028">Culture and climate
change: Japanese cherry blossom festivals and stakeholders<span>’</span>
knowledge and attitudes about global climate change</a>. <em>Biological
Conservation</em>, 144, 654–658.
</div>
<div id="ref-schölkopf1998" class="csl-entry" role="listitem">
Schölkopf, B., Smola, A. &amp; Müller, K.-R. (1998). <a href="https://doi.org/10.1162/089976698300017467">Nonlinear Component
Analysis as a Kernel Eigenvalue Problem</a>. <em>Neural
Computation</em>, 10, 1299–1319.
</div>
<div id="ref-shi2017" class="csl-entry" role="listitem">
Shi, P., Chen, Z., Reddy, G.V.P., Hui, C., Huang, J. &amp; Xiao, M.
(2017). <a href="https://doi.org/10.1016/j.agrformet.2017.04.001">Timing
of cherry tree blooming: Contrasting effects of rising winter low
temperatures and early spring temperatures</a>. <em>Agricultural and
Forest Meteorology</em>, 240-241, 78–89.
</div>
<div id="ref-smith2017" class="csl-entry" role="listitem">
Smith, M.L., Ruffley, M., Espíndola, A., Tank, D.C., Sullivan, J. &amp;
Carstens, B.C. (2017). <a href="https://doi.org/10.1111/mec.14223">Demographic model selection
using random forests and the site frequency spectrum</a>. <em>Molecular
Ecology</em>, 26, 4562–4573.
</div>
<div id="ref-søgaard2021" class="csl-entry" role="listitem">
Søgaard, A., Ebert, S., Bastings, J. &amp; Filippova, K. (2021). <a href="https://doi.org/10.18653/v1/2021.eacl-main.156">We need to talk
about random splits</a>. <em>Proceedings of the 16th Conference of the
European Chapter of the Association for Computational Linguistics: Main
Volume</em>.
</div>
<div id="ref-stock2023" class="csl-entry" role="listitem">
Stock, A., Gregr, E.J. &amp; Chan, K.M.A. (2023). <a href="https://doi.org/10.1038/s41559-023-02162-1">Data leakage
jeopardizes ecological applications of machine learning</a>. <em>Nature
Ecology &amp; Evolution</em>.
</div>
<div id="ref-strumbelj2013" class="csl-entry" role="listitem">
Štrumbelj, E. &amp; Kononenko, I. (2013). <a href="https://doi.org/10.1007/s10115-013-0679-x">Explaining prediction
models and individual predictions with feature contributions</a>.
<em>Knowledge and Information Systems</em>, 41, 647–665.
</div>
<div id="ref-sulmont2019" class="csl-entry" role="listitem">
Sulmont, E., Patitsas, E. &amp; Cooperstock, J.R. (2019). <a href="https://doi.org/10.1145/3287324.3287392">Can you teach me to
machine learn?</a> <em>Proceedings of the 50th ACM Technical Symposium
on Computer Science Education</em>.
</div>
<div id="ref-thorndike1953" class="csl-entry" role="listitem">
Thorndike, R.L. (1953). <a href="https://doi.org/10.1007/bf02289263">Who
belongs in the family?</a> <em>Psychometrika</em>, 18, 267–276.
</div>
<div id="ref-thuiller2004" class="csl-entry" role="listitem">
Thuiller, W., Araújo, M.B. &amp; Lavorel, S. (2004). <a href="https://doi.org/10.1046/j.0305-0270.2003.00991.x">Do we need
land<span>-</span>cover data to model species distributions in
Europe?</a> <em>Journal of Biogeography</em>, 31, 353–361.
</div>
<div id="ref-tipping1999" class="csl-entry" role="listitem">
Tipping, M.E. &amp; Bishop, C.M. (1999). <a href="https://doi.org/10.1111/1467-9868.00196">Probabilistic Principal
Component Analysis</a>. <em>Journal of the Royal Statistical Society
Series B: Statistical Methodology</em>, 61, 611–622.
</div>
<div id="ref-tredennick2021" class="csl-entry" role="listitem">
Tredennick, A.T., Hooker, G., Ellner, S.P. &amp; Adler, P.B. (2021). <a href="https://doi.org/10.1002/ecy.3336">A practical guide to selecting
models for exploration, inference, and prediction in ecology</a>.
<em>Ecology</em>, 102.
</div>
<div id="ref-tuia2022" class="csl-entry" role="listitem">
Tuia, D., Kellenberger, B., Beery, S., Costelloe, B.R., Zuffi, S.,
Risse, B., <em>et al.</em> (2022). <a href="https://doi.org/10.1038/s41467-022-27980-y">Perspectives in
machine learning for wildlife conservation</a>. <em>Nature
Communications</em>, 13, 792.
</div>
<div id="ref-Unal2017" class="csl-entry" role="listitem">
Unal, I. (2017). <a href="https://doi.org/10.1155/2017/3762651">Defining
an Optimal Cut-Point Value in ROC Analysis: An Alternative Approach</a>.
<em>Computational and Mathematical Methods in Medicine</em>, 2017, 1–14.
</div>
<div id="ref-unwin2021" class="csl-entry" role="listitem">
Unwin, A. &amp; Kleinman, K. (2021). <a href="https://doi.org/10.1111/1740-9713.01589">The Iris Data Set: In
Search of the Source of <span><em>Virginica</em></span></a>.
<em>Significance</em>, 18, 26–29.
</div>
<div id="ref-valavi2018" class="csl-entry" role="listitem">
Valavi, R., Elith, J., Lahoz-Monfort, J.J. &amp; Guillera-Arroita, G.
(2018). <a href="https://doi.org/10.1111/2041-210x.13107">block CV : An
r package for generating spatially or environmentally separated folds
for <span><em>k</em></span> <span>-</span>fold
cross<span>-</span>validation of species distribution models</a>.
<em>Methods in Ecology and Evolution</em>, 10, 225–232.
</div>
<div id="ref-vasseur2004" class="csl-entry" role="listitem">
Vasseur, D.A. &amp; Yodzis, P. (2004). <a href="https://doi.org/10.1890/02-3122">THE COLOR OF ENVIRONMENTAL
NOISE</a>. <em>Ecology</em>, 85, 1146–1152.
</div>
<div id="ref-vermote2016" class="csl-entry" role="listitem">
Vermote, E., Justice, C., Claverie, M. &amp; Franch, B. (2016). <a href="https://doi.org/10.1016/j.rse.2016.04.008">Preliminary analysis of
the performance of the Landsat 8/OLI land surface reflectance
product</a>. <em>Remote Sensing of Environment</em>, 185, 46–56.
</div>
<div id="ref-wadoux2023" class="csl-entry" role="listitem">
Wadoux, A.M.J.-C., Saby, N.P.A. &amp; Martin, M.P. (2023). <a href="https://doi.org/10.5194/soil-9-21-2023">Shapley values reveal the
drivers of soil organic carbon stock prediction</a>. <em>SOIL</em>, 9,
21–38.
</div>
<div id="ref-weiskopf2022" class="csl-entry" role="listitem">
Weiskopf, S.R., Harmáčková, Z.V., Johnson, C.G., Londoño-Murcia, M.C.,
Miller, B.W., Myers, B.J.E., <em>et al.</em> (2022). <a href="https://doi.org/10.1016/j.envsoft.2022.105318">Increasing the
uptake of ecological model results in policy decisions to improve
biodiversity outcomes</a>. <em>Environmental Modelling &amp;
Software</em>, 149, 105318.
</div>
<div id="ref-whittaker1962" class="csl-entry" role="listitem">
Whittaker, R.H. (1962). <a href="https://doi.org/10.1007/bf02860872">Classification of natural
communities</a>. <em>The Botanical Review</em>, 28, 1–239.
</div>
<div id="ref-whittingham2006" class="csl-entry" role="listitem">
WHITTINGHAM, M.J., STEPHENS, P.A., BRADBURY, R.B. &amp; FRECKLETON, R.P.
(2006). <a href="https://doi.org/10.1111/j.1365-2656.2006.01141.x">Why
do we still use stepwise modelling in ecology and behaviour?</a>
<em>Journal of Animal Ecology</em>, 75, 1182–1189.
</div>
<div id="ref-yau2015" class="csl-entry" role="listitem">
Yau, N. (2015). <a href="https://doi.org/10.1002/9781118722213">Visualize this</a>.
</div>
<div id="ref-youden1950" class="csl-entry" role="listitem">
Youden, W.J. (1950). <a href="https://doi.org/10.1002/1097-0142(1950)3:1<32::aid-cncr2820030106>3.0.co;2-3">Index
for rating diagnostic tests</a>. <em>Cancer</em>, 3, 32–35.
</div>
<div id="ref-zeng2000" class="csl-entry" role="listitem">
Zeng, X. &amp; Martinez, T.R. (2000). <a href="https://doi.org/10.1080/095281300146272">Distribution-balanced
stratified cross-validation for accuracy estimation</a>. <em>Journal of
Experimental &amp; Theoretical Artificial Intelligence</em>, 12, 1–12.
</div>
<div id="ref-zheng2021" class="csl-entry" role="listitem">
Zheng, Y., Zhou, Q., He, Y., Wang, C., Wang, X. &amp; Wang, H. (2021).
<a href="https://doi.org/10.3390/rs13040766">An Optimized Approach for
Extracting Urban Land Based on Log-Transformed DMSP-OLS Nighttime Light,
NDVI, and NDWI</a>. <em>Remote Sensing</em>, 13, 766.
</div>
<div id="ref-zhou2013" class="csl-entry" role="listitem">
Zhou, H. &amp; Jakobsson, E. (2013). <a href="https://doi.org/10.1371/journal.pone.0081100">Predicting
Protein-Protein Interaction by the Mirrortree Method: Possibilities and
Limitations</a>. <em>PLoS ONE</em>, 8, e81100.
</div>
</div>



</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    const typesetMath = (el) => {
      if (window.MathJax) {
        // MathJax Typeset
        window.MathJax.typeset([el]);
      } else if (window.katex) {
        // KaTeX Render
        var mathElements = el.getElementsByClassName("math");
        var macros = [];
        for (var i = 0; i < mathElements.length; i++) {
          var texText = mathElements[i].firstChild;
          if (mathElements[i].tagName == "SPAN") {
            window.katex.render(texText.data, mathElements[i], {
              displayMode: mathElements[i].classList.contains('display'),
              throwOnError: false,
              macros: macros,
              fleqn: false
            });
          }
        }
      }
    }
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        for (let i = 0; i < 2; i++) {
          container.appendChild(note.children[i].cloneNode(true));
        }
        typesetMath(container);
        return container.innerHTML
      } else {
        typesetMath(note);
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      typesetMath(note);
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>This work is licensed under the <a href="https://creativecommons.org/licenses/by-nc/4.0/">CC BY-NC 4.0 International License</a></p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>